pkg/thumbnail/storage/filesystem.go

```diff
@@ -74,14 +70,68 @@ func (s FileSystem) BuildKey(r Request) string {
 	filetype := r.Types[0]
 	filename := r.Resolution.String() + "." + filetype
 
-	key := new(bytes.Buffer)
-	key.WriteString(etag[:2])
-	key.WriteRune('/')
-	key.WriteString(etag[2:4])
-	key.WriteRune('/')
-	key.WriteString(etag[4:])
-	key.WriteRune('/')
-	key.WriteString(filename)
+	return filepath.Join(etag[:2], etag[2:4], etag[4:], filename)
+}
+
+func (s FileSystem) rootDir(key string) string {
+	p := strings.Split(key, string(os.PathSeparator))
+	return p[0]
+}
+
+func (s FileSystem) storeImage(key string, img []byte) (string, error) {
+	imgPath := filepath.Join(s.root, filesDir, key)
+	dir := filepath.Dir(imgPath)
+	if err := os.MkdirAll(dir, 0700); err != nil {
+		return "", errors.Wrapf(err, "error while creating directory %s", dir)
+	}
+
+	if _, err := os.Stat(imgPath); os.IsNotExist(err) {
+		f, err := os.Create(imgPath)
+		if err != nil {
+			return "", errors.Wrapf(err, "could not create file \"%s\"", key)
+		}
+		defer f.Close()
+
+		_, err = f.Write(img)
+		if err != nil {
+			return "", errors.Wrapf(err, "could not write to file \"%s\"", key)
+		}
+	}
+	return imgPath, nil
+}
+
+// userDir returns the path to the user directory.
+// The username is hashed before appending it on the path to prevent bugs caused by invalid folder names.
+// Also the hash is then splitted up in three parts that results in a path which looks as follows:
+// <filestorage-root>/users/<3 characters>/<3 characters>/<48 characters>/
+// This will balance the folders in setups with many users.
```

Comment:
nice! but what's the idea behind this? Avoiding collisions?

Reply:
The hashing should prevent invalid folder names like for windows  `CON, PRN, AUX, NUL`  or some special characters are also forbidden. We could also use the plain username if we provide a username validator somewhere like in ocis-pkg to be used in every service.

Reply:
ah, understand! thanks for explaining :D